BinaryEncoding/Decoding issues on arrays with two consecutive integers

<!-- Please answer these questions before submitting your issue. Thanks! -->

### What version of Genji are you using?

<pre>
$ genji version
Genji v0.11.0
Genji CLI (devel)
</pre>

### Does this issue reproduce with the latest release?

<!-- Yes/no -->

### What did you do?

```diff
~/code/src/github.com/genjidb/genji U main› git diff         
diff --git a/document/value_test.go b/document/value_test.go
index 181217f..721c601 100644
--- a/document/value_test.go
+++ b/document/value_test.go
@@ -386,6 +386,10 @@ func TestValueBinaryMarshaling(t *testing.T) {
 		{"double", document.NewDoubleValue(-3.14)},
 		{"text", document.NewTextValue("foo")},
 		{"blob", document.NewBlobValue([]byte("bar"))},
+		{"arrayfun", document.NewArrayValue(document.NewValueBuffer(
+			document.NewIntegerValue(2),
+			document.NewIntegerValue(98),
+		))},
 		{"array", document.NewArrayValue(document.NewValueBuffer(
 			document.NewBoolValue(true),
 			document.NewIntegerValue(55),
```

### What did you expect to see?

```
-*- mode: compilation; default-directory: "~/code/src/github.com/genjidb/genji/document/" -*-
Compilation started at Wed Mar  3 18:39:10

go test -run='TestValueBinaryMarshaling'
PASS
ok  	github.com/genjidb/genji/document	0.089s

Compilation finished at Wed Mar  3 18:39:11
```

### What did you see instead?

```
-*- mode: compilation; default-directory: "~/code/src/github.com/genjidb/genji/document/" -*-
Compilation started at Wed Mar  3 18:33:50

go test -run='TestValueBinaryMarshaling'
--- FAIL: TestValueBinaryMarshaling (0.00s)
    --- FAIL: TestValueBinaryMarshaling/arrayfun (0.00s)
        value_test.go:443: 
            	Error Trace:	value_test.go:443
            	Error:      	Received unexpected error:
            	            	malformed integer
            	Test:       	TestValueBinaryMarshaling/arrayfun
FAIL
exit status 1
FAIL	github.com/genjidb/genji/document	0.159s

Compilation exited abnormally with code 1 at Wed Mar  3 18:33:51
```

### What Go version and environment are you using?

<pre>
$ go version
go version go1.15.8 darwin/amd64

</pre>

<details><summary><code>go env</code> Output</summary><br><pre>
$ go env
GO111MODULE="on"
GOARCH="amd64"
GOBIN="/Users/tech/code/bin"
GOCACHE="/Users/tech/Library/Caches/go-build"
GOENV="/Users/tech/Library/Application Support/go/env"
GOEXE=""
GOFLAGS=""
GOHOSTARCH="amd64"
GOHOSTOS="darwin"
GOINSECURE=""
GOMODCACHE="/Users/tech/code/pkg/mod"
GONOPROXY=""
GONOSUMDB=""
GOOS="darwin"
GOPATH="/Users/tech/code"
GOPRIVATE=""
GOPROXY="https://proxy.golang.org,direct"
GOROOT="/Users/tech/.asdf/installs/golang/1.15.8/go"
GOSUMDB="sum.golang.org"
GOTMPDIR=""
GOTOOLDIR="/Users/tech/.asdf/installs/golang/1.15.8/go/pkg/tool/darwin_amd64"
GCCGO="gccgo"
AR="ar"
CC="clang"
CXX="clang++"
CGO_ENABLED="1"
GOMOD="/dev/null"
CGO_CFLAGS="-g -O2"
CGO_CPPFLAGS=""
CGO_CXXFLAGS="-g -O2"
CGO_FFLAGS="-g -O2"
CGO_LDFLAGS="-g -O2"
PKG_CONFIG="pkg-config"
GOGCCFLAGS="-fPIC -m64 -pthread -fno-caret-diagnostics -Qunused-arguments -fmessage-length=0 -fdebug-prefix-map=/var/folders/xl/dfsx6nzs4cb2y40m555mlfjh0000gn/T/go-build752488097=/tmp/go-build -gno-record-gcc-switches -fno-common"

</pre></details>

--- a/document/value_encoding.go
+++ b/document/value_encoding.go
@@ -218,7 +218,7 @@ func decodeValueUntil(data []byte, delim, end byte) (Value, int, error) {
 	case BoolValue:
 		i++
 	case IntegerValue, DoubleValue:
-		if i+8 < len(data) && data[i+8] == delim {
+		if i+8 < len(data) && (data[i+8] == delim || data[i+8] == end) {
 			i += 8
 		} else {
 			return Value{}, 0, errors.New("malformed " + t.String())

--- a/document/value_test.go
+++ b/document/value_test.go
@@ -389,13 +389,24 @@ func TestValueBinaryMarshaling(t *testing.T) {
 		{"array", document.NewArrayValue(document.NewValueBuffer(
 			document.NewBoolValue(true),
 			document.NewIntegerValue(55),
+			document.NewIntegerValue(56),
+			document.NewIntegerValue(57),
 			document.NewDoubleValue(789.58),
+			document.NewDoubleValue(790.58),
+			document.NewDoubleValue(791.58),
 			document.NewArrayValue(document.NewValueBuffer(
 				document.NewBoolValue(false),
 				document.NewIntegerValue(100),
 				document.NewTextValue("baz"),
 			)),
-			document.NewBlobValue([]byte("loo")),
+			document.NewArrayValue(document.NewValueBuffer(
+				document.NewBoolValue(true),
+				document.NewIntegerValue(101),
+				document.NewTextValue("bax"),
+			)),
+			document.NewBlobValue([]byte("coc")),
+			document.NewBlobValue([]byte("ori")),
+			document.NewBlobValue([]byte("co!")),
 			document.NewDocumentValue(
 				document.NewFieldBuffer().
 					Add("foo1", document.NewBoolValue(true)).

